## 동기와 비동기

- "동기(Sync)로 실행된다" = 먼저 실행된 코드의 결과가 나올때까지 대기함
    
- "비동기(Async)로 실행된다" = 실행된 순서와 관계 없이 결과가 나옴
   
  자바 스크립트는 비동기식으로 실행이 되기때문에 first가 맨 마지막에 출력된느 것을 확인할 수 있다.


```javascript
function first() {
  console.log('First');
}

setTimeout(first, 1000); // 1000ms(1초) 뒤에 first 함수를 실행해준다.

console.log('Middle');
console.log('Last');
```

    Middle
    Last
    First


- Blocking Model에서는 코드의 실행이 끝나기 전까지 실행 제어권을 다른곳에 넘기지 않아 다른 작업을 하지 못하고 대기
- Non-blocking Model에서는 코드의 실행이 끝나지 않아도 실행 제어권을 다른곳에 넘겨 다음 코드가 실행.

즉, 제어권을 넘기면(Non-blocking) 다른 코드도 실행될 수 있으므로 비동기 처리가 가능하지만 제어권을 넘기지 않으면(Blocking) 비동기 처리가 가능한 환경이어도 비동기 처리가 불가.

- 자바스크립트는 Async + Non-blocking Model을 채용하여 현재 실행중인 코드의 실행이 끝나지 않아도 다음 코드를 호출
- 결론적으로 자바스크립트는 Non-blocking model을 가지며 각 명령이 순서대로 실행될 수 있게 구현되어 있지만, Non-blocking model에 의해 명령이 아닌 모든 함수는 비동기적으로 실행.

## Promise

- 자바스크립트에서 비동기 처리를 동기로 처리할 수 있게 돕는 객체
이 객체를 이용하면 여러분은 Non-blocking model을 가진 자바스크립트에서도 비동기 처리에 용이
- Promise 생성자
`executor`에는 함수만 올 수 있으며 인자로 resolve, reject가 주입
    
    ```jsx
    new Promise(executor);
    
    // 예제
    new Promise((resolve, reject) => {
    	// 명령문
    });
    ```
    
- 프로미스의 상태
    
    프로미스는 반드시 3가지 상태를 지니며, 대기(Pending) 상태가 아니라면 Promise의 연산이 이미 끝난 상태로 볼 수 있습니다.
    
    - 대기(*Pending)*: 이행하거나 거부되지 않은 초기 상태.
    - 이행(*Fulfilled)*: 연산이 성공적으로 완료됨.  /resolve
    - 거부(*Rejected)*: 연산이 실패함. / reject


    ```
    const timerPromise = new Promise((resolve, reject) => { // 이곳에 정의된 함수가 executor
      setTimeout(() => {
    	  console.log('First');
    		resolve();
    	}, 1000);
    });
    
    // 이 시점에서 timerPromise는 Fulfilled Promise라고 부를 수 있다.
    
    timerPromise.then(() => {
    	console.log('Middle');
    	console.log('Last');
    });
    
    // Print: First
    // Middle
    // Last
    ```
    
- Promise.catch
    
    프로미스 안에서 에러가 throw 되거나 reject되면 catch 메서드에 작성한 함수가 실행.
    
    ```jsx
    const errorPromise = new Promise((resolve, reject) => {
      setTimeout(() => {
    	  console.log('First');
    		reject('Error!!'); // 직접 reject를 실행하면 프로미스에서 에러가 발생한것으로 간주됩니다.
    	}, 1000);
    });
    
    errorPromise.then(() => {
    	console.log('Middle');
    	console.log('Last');
    }).catch((error) => {
    	console.log('에러 발생!', error);
    });
    
    // Print: '에러 발생! Error!!'
    ```
   

### await

- await 연산자를 사용하면 Promise가 fulfill되거나 rejected될 때 까지 함수의 실행을 중단하고 기다.
Promise의 연산이 끝나면 함수에서 반환한 값을 얻을 수 있습니다.
- **await 연산자는 현재, async 함수 안에서만 사용할 수 있음.**
- 인터페이스
        const result = await 값(promise가 아니어도 정상작동)


```javascript
timerPromise1 = new Promise((resolve, reject) => { // 이곳에 정의된 함수가 executor
  setTimeout(() => {
      console.log('First');
        resolve();
    }, 1000);
});

// 이 시점에서 timerPromise는 Fulfilled Promise라고 부를 수 있다.

timerPromise1.then(()=>{
    console.log("Second");
});

```




    Promise { <pending> }



    First
    Second



```javascript
timerPromise1 = new Promise((resolve, reject) => { // 이곳에 정의된 함수가 executor
  setTimeout(() => {
      console.log('First');
        resolve();
    }, 1000);
});

async function timer2() {
    const result = timerPromise1;
    console.log("Second");
}

timer2();

```

    Second





    Promise { undefined }



    First



```javascript
timerPromise1 = new Promise((resolve, reject) => { // 이곳에 정의된 함수가 executor
  setTimeout(() => {
      console.log('First');
        resolve();
    }, 1000);
});

async function timer2() {
    const result = await timerPromise1;
    console.log("Second");
}

timer2();

```




    Promise { <pending> }



    First
    Second

